tests/Functional/Platform/AggregateConcatExpressionTest.php

```diff
@@ -0,0 +1,34 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\DBAL\Tests\Functional\Platform;
+
+use Doctrine\DBAL\ParameterType;
+use Doctrine\DBAL\Schema\Table;
+use Doctrine\DBAL\Tests\FunctionalTestCase;
+use function explode;
+use function sprintf;
+
+class AggregateConcatExpressionTest extends FunctionalTestCase
+{
+    public function testAggregateConcat() : void
+    {
+        $table = new Table('aggregate_concat_test');
+        $table->addColumn('value', 'string', ['length' => 10]);
+        $this->connection->getSchemaManager()->dropAndCreateTable($table);
+        $this->connection->insert('aggregate_concat_test', ['value' => 'foo'], ['value' => ParameterType::STRING]);
+        $this->connection->insert('aggregate_concat_test', ['value' => 'bar'], ['value' => ParameterType::STRING]);
+        $this->connection->insert('aggregate_concat_test', ['value' => 'baz'], ['value' => ParameterType::STRING]);
+
+        $platform = $this->connection->getDatabasePlatform();
+
+        $sql = sprintf('SELECT %s FROM aggregate_concat_test', $platform->getAggregateConcatExpression('value', $this->connection->quote(',')));
+
+        // The order is not guaranteed, so we can only check if the parts are the same
```

Comment:
Is it not guaranteed on a specific platform or it's by design? This way the abstracted out function is barely usable, and when the order is important (which is often the case), the consumer will have to resort to using raw SQL.

Reply:
Unfortunately not all platforms support an order (SQLite being at least one of them). I figured that passing an order as an extra optional parameter could be a possible future extension of this method. That would mean that a platform like SQLite would have to throw `NotSupported` if that parameter is set, though.

While some use cases definitely require ordering, there are also cases where having no ordering is fine, or sorting can happen afterwards. E.g. when you have tables `posts` and `tags` with a many-to-many relationship between them, concat aggregation could give you all posts with all their tags in one single database query. Otherwise such an operation would require multiple queries, or doing something fetch-join-like and manual deduplication of rows.

Reply:
In your example with posts and tags, using the string aggregation function will only work if only the tag names are needed. If anything else than a name is needed, this query will have to be reworked to using two queries: one retrieves all posts and their identifiers, the other retrieves all tag information by the identifiers collected by the first query.

> Otherwise such an operation would require multiple queries, or doing something fetch-join-like and manual deduplication of rows.

Performance-wise, those two simple queries will be most likely faster than the one with the join and group by.

While I don't have a strong objection, it looks like the API proposed helps solve a problem in a way that is not extensible and will require a rework w/o using this API if needed.

Reply:
Another example is, let's say each post has more than one author and more than one tag. As long as you're only showing tags in the list, you can string-aggregate them but as soon as you also need to show the authors, you'll have to add a secondary query.

Reply:
I could try to add an optional `$orderBy` parameter to this method if you think it would make this feature much more useful. That would mean it must throw `NotSupported` if `$orderBy` is set and the platform does not support an order, though.

> Another example is, let's say each post has more than one author and more than one tag. As long as you're only showing tags in the list, you can string-aggregate them but as soon as you also need to show the authors, you'll have to add a secondary query.

I agree that concat-aggregation surely has its limitations. I just checked our codebase, and we only use it in a few select places. Still, given that all major DB platforms support it, it seems useful to have an abstraction for in the DBAL platform.

Reply:
Some research showed that all platforms except SQLite support ordering, so I just added an optional `$orderBy` parameter.